From 79e65c43a56da12fe236881dff4245b20fbdb494 Mon Sep 17 00:00:00 2001
From: Jim Somerville <Jim.Somerville@windriver.com>
Date: Fri, 10 Apr 2026 12:58:06 -0300
Subject: [PATCH] STX: migration thread affinity and priority qmp

This includes de-blacklisting the scheduler and affinity
setting syscalls.

Files moved from softmmu/ to system/ in qemu 10.0.8:
- softmmu/qemu-seccomp.c -> system/qemu-seccomp.c
- softmmu/trace-events -> system/trace-events

Removed DEBUG/DPRINTF block from the original patch as qemu 10.0.8
uses the trace subsystem for debug logging (trace event
migrate_thread added in system/trace-events).

Added guards around sched_setaffinity and sched_setscheduler calls
in migration_thread to only execute when the user has explicitly
configured affinity or priority via QMP. Without these guards,
sched_setaffinity with an empty mask (0) returns EINVAL, and
sched_setscheduler with SCHED_FIFO priority 0 also returns EINVAL
(valid range is 1-99 [1]), causing spurious error messages on
every migration where these options are not configured.

[1] https://man7.org/linux/man-pages/man2/sched_setscheduler.2.html

Signed-off-by: Jim Somerville <Jim.Somerville@windriver.com>
[ Update hmp-commands struc ]
Signed-off-by: Rafael Falcao <Rafael.VieiraFalcao@windriver.com>
[Rebased original changes to the qemu 7.2 version]
Signed-off-by: david.liu <david.liu@windriver.com>
[Rebased to qemu 10.0.8 for Trixie]
Signed-off-by: cbohlhal <Caique.BohlhalterdeSouza@windriver.com>
---
 hmp-commands.hx       | 30 +++++++++++++++++++++++++++++
 include/monitor/hmp.h |  2 ++
 migration/migration.c | 48 ++++++++++++++++++++++++++++++++++++++++++++++
 monitor/hmp-cmds.c    | 26 +++++++++++++++++++++++++
 qapi/misc.json        | 26 +++++++++++++++++++++++++
 system/qemu-seccomp.c |  5 -----
 6 files changed, 132 insertions(+), 5 deletions(-)

diff --git a/hmp-commands.hx b/hmp-commands.hx
index 06746f0afc..7e2b107208 100644
--- a/hmp-commands.hx
+++ b/hmp-commands.hx
@@ -39,6 +39,36 @@ SRST
   completes.
 ERST
 
+    {
+        .name       = "migrate-set-thread-cpumask",
+        .args_type  = "value:o",
+        .params     = "value",
+        .help       = "Set CPU mask for the migration thread."
+                      "Defaults to CPU 0 if no mask is specified",
+        .cmd        = hmp_migrate_set_thread_cpumask,
+   },
+
+SRST
+``migrate_set_thread_cpumask`` *value*
+  Set CPU mask for the migration thread
+ERST
+
+    {
+        .name       = "migrate-set-thread-priority",
+        .args_type  = "value:o",
+        .params     = "value",
+        .help       = "Set real time priority for the migration thread."
+                      "Defaults to no change migration thread priority if not"
+                      "specified or out of range. Range [1-99].Scheduling"
+                      "policy will always be- SCHED_FIFO",
+        .cmd        = hmp_migrate_set_thread_priority,
+    },
+
+SRST
+``migrate_set_thread_priority`` *value*
+  Set real time priority for the migration thread
+ERST
+
     {
         .name       = "quit|q",
         .args_type  = "",
diff --git a/include/monitor/hmp.h b/include/monitor/hmp.h
index ae116d9804..e64b0d56d3 100644
--- a/include/monitor/hmp.h
+++ b/include/monitor/hmp.h
@@ -70,6 +70,8 @@ void hmp_migrate_set_parameter(Monitor *mon, const QDict *qdict);
 void hmp_client_migrate_info(Monitor *mon, const QDict *qdict);
 void hmp_migrate_start_postcopy(Monitor *mon, const QDict *qdict);
 void hmp_x_colo_lost_heartbeat(Monitor *mon, const QDict *qdict);
+void hmp_migrate_set_thread_cpumask(Monitor *mon, const QDict *qdict);
+void hmp_migrate_set_thread_priority(Monitor *mon, const QDict *qdict);
 void hmp_set_password(Monitor *mon, const QDict *qdict);
 void hmp_expire_password(Monitor *mon, const QDict *qdict);
 void hmp_change(Monitor *mon, const QDict *qdict);
diff --git a/migration/migration.c b/migration/migration.c
index 12699bf165..6731d81fca 100644
--- a/migration/migration.c
+++ b/migration/migration.c
@@ -43,6 +43,7 @@
 #include "qapi/qapi-visit-migration.h"
 #include "qapi/qapi-visit-sockets.h"
 #include "qapi/qapi-commands-migration.h"
+#include "qapi/qapi-commands-misc.h"
 #include "qapi/qapi-events-migration.h"
 #include "qapi/qmp/qerror.h"
 #include "qobject/qnull.h"
@@ -64,6 +65,7 @@
 #include "system/cpus.h"
 #include "yank_functions.h"
 #include "system/qtest.h"
+#include <sched.h>
 #include "options.h"
 #include "system/dirtylimit.h"
 #include "qemu/sockets.h"
@@ -74,6 +76,11 @@
 
 #define INMIGRATE_DEFAULT_EXIT_ON_ERROR true
 
+/* variables for pinning the migration thread to a CPU and assigning the
+* realtime priority to it */
+static uint64_t migrate_thread_cpumask=0;
+static uint64_t migrate_thread_priority=0;
+
 static NotifierWithReturnList migration_state_notifiers[] = {
     NOTIFIER_ELEM_INIT(migration_state_notifiers, MIG_MODE_NORMAL),
     NOTIFIER_ELEM_INIT(migration_state_notifiers, MIG_MODE_CPR_REBOOT),
@@ -1871,6 +1878,30 @@ int migrate_add_blocker_modes(Error **reasonp, Error **errp, MigMode mode, ...)
     return add_blockers(reasonp, errp, modes);
 }
 
+void qmp_migrate_set_thread_cpumask(int64_t value, Error **errp)
+{
+    /* Check for truncation */
+    if (value != (size_t)value) {
+        error_setg(errp, "Migration thread CPU Mask exceeding address space");
+        return;
+    }
+    /*resize the value */
+    value >>= 20; /*Magic */
+    migrate_thread_cpumask = value;
+}
+
+void qmp_migrate_set_thread_priority(int64_t value, Error **errp)
+{
+    /* Check for truncation */
+    if (value != (size_t)value) {
+        error_setg(errp, "Migration thread Priority exceeding address space");
+        return;
+    }
+    /*resize the value */
+    value >>= 20;
+    migrate_thread_priority = value;
+}
+
 int migrate_add_blocker_internal(Error **reasonp, Error **errp)
 {
     int modes = BIT(MIG_MODE__MAX) - 1;
@@ -3677,6 +3708,23 @@ static void *migration_thread(void *opaque)
         qemu_savevm_send_postcopy_advise(s->to_dst_file);
     }
 
+    /* Bind Migration thread to the processor specified by the user */
+    if (migrate_thread_cpumask != 0) {
+        if (sched_setaffinity(0, sizeof(migrate_thread_cpumask), (cpu_set_t *)&migrate_thread_cpumask) < 0) {
+            error_report("migration: failed to set thread affinity");
+        }
+    }
+
+    /* Change the realtime priority of the migration thread specified by the user */
+    if (migrate_thread_priority > 0) {
+        struct sched_param schedp;
+        memset(&schedp, 0, sizeof(schedp));
+        schedp.sched_priority = migrate_thread_priority;
+        if (sched_setscheduler(0, SCHED_FIFO, &schedp) < 0) {
+            error_report("migration: failed to set thread priority");
+        }
+    }
+
     if (migrate_colo()) {
         /* Notify migration destination that we enable COLO */
         qemu_savevm_send_colo_enable(s->to_dst_file);
diff --git a/monitor/hmp-cmds.c b/monitor/hmp-cmds.c
index 7ded3378cf..f11d05634e 100644
--- a/monitor/hmp-cmds.c
+++ b/monitor/hmp-cmds.c
@@ -146,6 +146,32 @@ void hmp_cont(Monitor *mon, const QDict *qdict)
     hmp_handle_error(mon, err);
 }
 
+void hmp_migrate_set_thread_cpumask(Monitor *mon, const QDict *qdict)
+{
+    int64_t value = qdict_get_int(qdict, "value");
+    Error *err = NULL;
+
+    qmp_migrate_set_thread_cpumask(value, &err);
+    if (err) {
+        monitor_printf(mon, "%s\n", error_get_pretty(err));
+        error_free(err);
+        return;
+    }
+}
+
+void hmp_migrate_set_thread_priority(Monitor *mon, const QDict *qdict)
+{
+    int64_t value = qdict_get_int(qdict, "value");
+    Error *err = NULL;
+
+    qmp_migrate_set_thread_priority(value, &err);
+    if (err) {
+        monitor_printf(mon, "%s\n", error_get_pretty(err));
+        error_free(err);
+        return;
+    }
+}
+
 void hmp_change(Monitor *mon, const QDict *qdict)
 {
     const char *device = qdict_get_str(qdict, "device");
diff --git a/qapi/misc.json b/qapi/misc.json
index 559b66f201..5f6b6d6d21 100644
--- a/qapi/misc.json
+++ b/qapi/misc.json
@@ -243,6 +243,32 @@
   'returns': 'str',
   'features': [ 'savevm-monitor-nodes' ] }
 
+##
+# @migrate-set-thread-cpumask:
+#
+# Set migration thread CPU mask.
+#
+# @value: CPU mask.
+#
+# .. note:: A value lesser than zero will be automatically round up to zero.
+#
+# Since: 0.14.0
+##
+{ 'command': 'migrate-set-thread-cpumask', 'data': {'value': 'int'} }
+
+##
+# @migrate-set-thread-priority:
+#
+# Set migration thread Real Timer priority mask.
+#
+# @value: Thread Priority.
+#
+# .. note:: A value lesser than zero will be automatically round up to zero.
+#
+# Since: 0.14.0
+##
+{ 'command': 'migrate-set-thread-priority', 'data': {'value': 'int'} }
+
 ##
 # @getfd:
 #
diff --git a/system/qemu-seccomp.c b/system/qemu-seccomp.c
index f8e1238b91..073e0ac6da 100644
--- a/system/qemu-seccomp.c
+++ b/system/qemu-seccomp.c
@@ -258,11 +258,6 @@ static const struct QemuSeccompSyscall denylist[] = {
       0, NULL, SCMP_ACT_ERRNO(EPERM) },
     { SCMP_SYS(sched_setparam),         QEMU_SECCOMP_SET_RESOURCECTL,
       0, NULL, SCMP_ACT_ERRNO(EPERM) },
-    { SCMP_SYS(sched_setscheduler),     QEMU_SECCOMP_SET_RESOURCECTL,
-      ARRAY_SIZE(sched_setscheduler_arg), sched_setscheduler_arg,
-      SCMP_ACT_ERRNO(EPERM) },
-    { SCMP_SYS(sched_setaffinity),      QEMU_SECCOMP_SET_RESOURCECTL,
-      0, NULL, SCMP_ACT_ERRNO(EPERM) },
 };
 
 static inline __attribute__((unused)) int
-- 
2.34.1

